crates/compiler/can/src/desugar.rs

```diff
@@ -840,33 +877,670 @@ pub fn desugar_expr<'a>(
+                    }
+                }
+
+                if let Some(ValueDef::Stmt(_)) = type_or_value_def.err() {
+                    internal_error!("handle Stmt");
+                }
+            }
+
+            // try to unwrap the loc_ret
+            match unwrap_suffixed_expression(arena, loc_ret, false, module_path) {
+                Ok(new_loc_ret) => {
+                    Ok(arena.alloc(Loc::at(loc_expr.region, Defs(arena.alloc(local_defs), new_loc_ret))))
+                },
+                Err(EUnwrapped::UnwrappedExpr(new_loc_ret)) => {
+                    // the loc_ret was unwrapped, replace and return new expression
+                    Err(EUnwrapped::UnwrappedExpr(arena.alloc(Loc::at(loc_expr.region, Defs(defs, new_loc_ret)))))
+                }
+                Err(EUnwrapped::UnwrappedSubExpr { arg, pat, new }) => {
+                    // the loc_ret was unwrapped
+                    let new_loc_ret = apply_task_await(arena, loc_ret.region, arg, pat, new);
+                    Err(EUnwrapped::UnwrappedExpr(arena.alloc(Loc::at(loc_expr.region, Defs(defs, new_loc_ret)))))
+                },
+            }
+        },
+        _ => internal_error!("unreachable, expected a Defs node to be passed into unwrap_suffixed_expression_defs_help"),
+    }
+}
+
+#[cfg(test)]
+mod unwrap_suffixed_expression_tests {
+
+    use crate::desugar::desugar_defs_node_values;
+    use crate::desugar::Bump;
+    use roc_parse::test_helpers::parse_defs_with;
+    use roc_test_utils::assert_multiline_str_eq;
+
+    #[test]
+    fn simple_pizza() {
+        /*
+        "hello" |> line!
+
+        Task.ok {}
+
+        # desugared
+        Task.await ("hello" |> line) \{} -> Task.ok {}
+        */
+
+        let arena = &Bump::new();
+
+        let src = r#"
+            main = 
+                "hello" |> line!
+
+                Task.ok {}
+            "#;
+
+        let mut defs = parse_defs_with(arena, src).unwrap();
+
+        desugar_defs_node_values(arena, &mut defs, src, &mut None, "test.roc", true);
+
+        let expected = r##"Defs { tags: [Index(2147483648)], regions: [@0-40], space_before: [Slice(start = 0, length = 0)], space_after: [Slice(start = 0, length = 0)], spaces: [], type_defs: [], value_defs: [Body(@0-4 Identifier { ident: "main", suffixed: 0 }, @24-40 Apply(@24-40 Var { module_name: "Task", ident: "await", suffixed: 0 }, [@24-40 Apply(@24-40 Var { module_name: "", ident: "line", suffixed: 0 }, [@24-31 Str(PlainLine("hello"))], BinOp(Pizza)), @24-40 Closure([@24-40 RecordDestructure([])], @58-68 Apply(@58-65 Var { module_name: "Task", ident: "ok", suffixed: 0 }, [@66-68 Record([])], Space))], BangSuffix))] }"##;
+
+        assert_multiline_str_eq!(format!("{:?}", &defs).as_str(), expected);
+    }
+
+    #[test]
+    fn single_suffix() {
+        /*
+        ## Example with single suffix
+        x = foo!
+        bar x
+
+        # desugared
+        Task.await (foo) \x -> bar x
+        */
```

Comment:
I'm quite happy with this testing approach.. It makes it much easier to follow the unwrapping of suffixed expressions and confirm they are correct.